![](images/ikmb_bfx_logo.png)

# LaMeta

## UNDER CONSTRUCTION

## Overview

This pipeline takes metagenomic (paired-end) short-read data as input, as generated by Illumina sequencing. From this data, the pipeline aims to assemble high-quality single genomes.

1. All samples are separately quality controlled to remove Illumina library adaptors, low quality sequences and sequence ends, and possible host-genetic (human) and spike in contamination (PhiX). The cleaned reads are used for the subsequently following steps.
2. For all samples a separate metagenomic assembly is performed using Spades in metagenomic mode. The sequences are then mapped back to the resulting scaffolds and binned using the MaxBin2 software.
3. Additionally, a co-assembly with Megahit is performed. Using a groupfile it is possible to split samples into separate groups for this co-assembly. Again the resulting contigs are used as reference for backmapping, followed by two separate binning approaches using MaxBin2 and Metabat2, which for this approach now can also incorporate across-sample abundance differences for the binning procedure.
4. The resulting bins from the single-sample and subgroup co-assembly approaches are finally dereplicated using the dRep package, to achieve the highest-possible quality of single-genome bins combined with low redundancy.
5. All samples are again mapped to the final resulting bins to estimate bin abundance.


## Executing the pipeline

To execute the pipeline with all default settings, do:

`nextflow -c nextflow.config run main.nf --reads '/path/to/*_R{1,2}_001.fastq.gz' --host_index /path/to/index`

Where `host_index` is a folder containing a BBMap index directory ("ref"). 

Alternatively, the pipeline can also create an index for your, if you provide a FASTA file:

`nextflow -c nextflow.config run main.nf --reads '/path/to/*_R{1,2}_001.fastq.gz' --host /path/to/genome.fa`

For details on how to prepare your reference database, please also see [here](http://seqanswers.com/forums/showthread.php?t=42552).

### Optional parameters

The pipeline uses several parameters to fine-tune the various pipeline stages. Some of these can be modified during pipeline execution.

For a list of all parameters, see:

`nextflow -c nextflow.config run main.nf --help`

## Reading the output

## Dependencies and versions
 * [BBMap (v.37.88)](https://sourceforge.net/projects/bbmap/): QC, Mapping to contigs.
 * [Megahit (v1.1.2)](https://github.com/voutcn/megahit): Groupwise Co-Assemblies.
 * [Spades (v.3.9.0)](http://bioinf.spbau.ru/en/spades): Single-Sample Assemblies.
 * [Samtools (v.1.5)](http://www.htslib.org): Conversion of SAM to BAM.
 * [MaxBin2 (v.2.2.4)](https://downloads.jbei.org/data/microbial_communities/MaxBin/MaxBin.html): Binning of Contigs.
 * [Metabat2 (v.2.12.1)](https://bitbucket.org/berkeleylab/metabat/overview): Binning of Contigs.
 * [dRep (v.2.0.5)](https://github.com/MrOlm/drep): Evaluation of Binned Contigs.
 * [CheckM (v.1.0.11)](https://ecogenomics.github.io/CheckM/): Used by dRep.

## Code structure

This pipeline is comprised of the following components:

* main.nf (the actual workflow definition)
* nextflow.config (the top-level configuration file with generic options)
* config/ (compute environment-specific configuration options)
* README.md (this file)

## Credit

This pipeline was developed by M. Rühlemann and M. Höppner. 
